api/services.go

```diff
@@ -0,0 +1,92 @@
+package api
+
+import (
+	"fmt"
+	"time"
+)
+
+// ServiceCheck represents the consul health check that Nomad registers.
+type ServiceCheck struct {
+	//FIXME Id is unused. Remove?
+	Id            string
+	Name          string
+	Type          string
+	Command       string
+	Args          []string
+	Path          string
+	Protocol      string
+	PortLabel     string `mapstructure:"port"`
+	AddressMode   string `mapstructure:"address_mode"`
+	Interval      time.Duration
+	Timeout       time.Duration
+	InitialStatus string `mapstructure:"initial_status"`
+	TLSSkipVerify bool   `mapstructure:"tls_skip_verify"`
+	Header        map[string][]string
+	Method        string
+	CheckRestart  *CheckRestart `mapstructure:"check_restart"`
+	GRPCService   string        `mapstructure:"grpc_service"`
+	GRPCUseTLS    bool          `mapstructure:"grpc_use_tls"`
+}
+
+// Service represents a Consul service definition.
+type Service struct {
+	//FIXME Id is unused. Remove?
+	Id           string
+	Name         string
+	Tags         []string
+	CanaryTags   []string `mapstructure:"canary_tags"`
+	PortLabel    string   `mapstructure:"port"`
+	AddressMode  string   `mapstructure:"address_mode"`
+	Checks       []ServiceCheck
+	CheckRestart *CheckRestart `mapstructure:"check_restart"`
+	Connect      *ConsulConnect
+}
+
+// Canonicalize the Service by ensuring its name and address mode are set. Task
+// will be nil for group services.
+func (s *Service) Canonicalize(t *Task, tg *TaskGroup, job *Job) {
+	if s.Name == "" {
+		if t != nil {
+			s.Name = fmt.Sprintf("%s-%s-%s", *job.Name, *tg.Name, t.Name)
+		} else {
+			s.Name = fmt.Sprintf("%s-%s", *job.Name, *tg.Name)
+		}
+	}
+
+	// Default to AddressModeAuto
+	if s.AddressMode == "" {
+		s.AddressMode = "auto"
+	}
+
+	// Canonicalize CheckRestart on Checks and merge Service.CheckRestart
+	// into each check.
+	for i, check := range s.Checks {
+		s.Checks[i].CheckRestart = s.CheckRestart.Merge(check.CheckRestart)
```

Comment:
Great question! Our usual pattern is to make Merge & Canonicalize return nil if called on a nil struct. A bit of an unusual pattern but because those methods are usually being called on user submitted data, it's nice to have the extra defensiveness always there: https://github.com/hashicorp/nomad/blob/v0.9.4/api/tasks.go#L287-L289

You did make me realize I need to move CheckRestart into this file too!